datafusion/core/tests/sqllogictests/src/main.rs

```diff
@@ -15,112 +15,135 @@
-    ///
-    /// The default implementation is `std::thread::sleep`, which is universial to any async runtime
-    /// but would block the current thread. If you are running in tokio runtime, you should override
-    /// this by `tokio::time::sleep`.
-    async fn sleep(dur: Duration) {
-        tokio::time::sleep(dur).await;
-    }
-}
+const TEST_DIRECTORY: &str = "tests/sqllogictests/test_files";
+const TEST_DIRECTORY_POSTGRES: &str = "tests/sqllogictests/postgres/test_files";
 
 #[tokio::main]
 #[cfg(target_family = "windows")]
-pub async fn main() -> Result<()> {
+pub async fn main() -> Result<(), Box<dyn Error>> {
     println!("Skipping test on windows");
     Ok(())
 }
 
 #[tokio::main]
 #[cfg(not(target_family = "windows"))]
-pub async fn main() -> Result<()> {
+pub async fn main() -> Result<(), Box<dyn Error>> {
     // Enable logging (e.g. set RUST_LOG=debug to see debug logs)
-
-    use sqllogictest::{default_validator, update_test_file};
     env_logger::init();
 
     let options = Options::new();
 
-    // default to all files in test directory filtering based on name
-    let files: Vec<_> = std::fs::read_dir(TEST_DIRECTORY)
-        .unwrap()
-        .map(|path| path.unwrap().path())
-        .filter(|path| options.check_test_file(path.as_path()))
-        .collect();
+    let files: Vec<_> = read_test_files(&options);
 
     info!("Running test files {:?}", files);
 
     for path in files {
-        println!("Running: {}", path.display());
-
         let file_name = path.file_name().unwrap().to_str().unwrap().to_string();
 
-        // Create the test runner
-        let ctx = context_for_test_file(&file_name).await;
-        let mut runner = sqllogictest::Runner::new(DataFusion { ctx, file_name });
-
-        // run each file using its own new DB
-        //
-        // We could run these tests in parallel eventually if we wanted.
         if options.complete_mode {
-            info!("Using complete mode to complete {}", path.display());
-            let col_separator = " ";
-            let validator = default_validator;
-            update_test_file(path, runner, col_separator, validator)
-                .await
-                .map_err(|e| e.to_string())?;
+            run_complete_file(&path, file_name).await?;
+        } else if options.postgres_mode {
+            run_postgres_test_file(&path, file_name).await?;
         } else {
-            // run the test normally:
-            runner.run_file_async(path).await?;
+            run_test_file(&path, file_name).await?;
         }
     }
 
     Ok(())
 }
 
+async fn run_test_file(path: &PathBuf, file_name: String) -> Result<(), Box<dyn Error>> {
+    println!("Running: {}", path.display());
+    let ctx = context_for_test_file(&file_name).await;
+    let mut runner = sqllogictest::Runner::new(DataFusion::new(ctx, file_name));
+    runner.run_file_async(path).await?;
+    Ok(())
+}
+
+async fn run_postgres_test_file(
```

Comment:
This function is the only new code here.
It is quite simple:
- copies `.slt` file to a temp location
- updates the temporary `.slt` file with Postgres results
- runs the temporary `.slt` with predefined results file with Datafusion

This way expected results were produced by Postgres. Thus running a test with Datafusion compares results of two engines.

Reply:
I think I would make more sense if the expected results were  part of the .slt file, as they are in other slt based systems, rather than dynamically generating the content.

Reply:
So, would you prefer to have explicit test cases like this one

```
query R
select 2.0;
----
2.0
```

instead of

```
query R
select 2.0;
```

Those who create new tests would use the `--complete` functionality to fill in the gaps.

I can implement comparison this way, of course. Now the `--complete` is called implicitly during a test run.

Reply:
I updated the tests to have explicit results.